unit_tests/test_cepch_iscsi_charm.py

```diff
@@ -0,0 +1,321 @@
+        self.patch_all()
+
+    def patch(self, method):
+        _m = patch.object(self.obj, method)
+        mock = _m.start()
+        self.addCleanup(_m.stop)
+        return mock
+
+    def patch_all(self):
+        for method in self.patches:
+            setattr(self, method, self.patch(method))
+
+
+class TestCephISCSIGatewayCharmBase(CharmTestCase):
+
+    PATCHES = [
+        'ch_templating',
+        'gwcli_client',
+        'subprocess',
+    ]
+
+    def setUp(self):
+        super().setUp(charm, self.PATCHES)
+        self.harness = Harness(
+            charm.CephISCSIGatewayCharmBase,
+        )
+        self.gwc = MagicMock()
+        self.gwcli_client.GatewayClient.return_value = self.gwc
+
+        # BEGIN: Workaround until
+        # https://github.com/canonical/operator/pull/196 lands
+        class _TestingOPSModelBackend(_TestingModelBackend):
+
+            def relation_ids(self, relation_name):
+                return self._relation_ids_map.get(relation_name, [])
+
+            # Hardcoded until network_get is implemented in
+            # _TestingModelBackend
+            def network_get(self, endpoint_name, relation_id=None):
+                network_data = {
+                    'bind-addresses': [{
+                        'interface-name': 'eth0',
+                        'addresses': [{
+                            'cidr': '10.0.0.0/24',
+                            'value': '10.0.0.10'}]}],
+                    'ingress-addresses': ['10.0.0.10'],
+                    'egress-subnets': ['10.0.0.0/24']}
+                return network_data
+
+        self.harness._backend = _TestingOPSModelBackend(
+            self.harness._unit_name)
+        self.harness._model = model.Model(
+            self.harness._unit_name,
+            self.harness._meta,
+            self.harness._backend)
+        self.harness._framework = framework.Framework(
+            ":memory:",
+            self.harness._charm_dir,
+            self.harness._meta,
+            self.harness._model)
+        # END Workaround
+
+    def test_init(self):
+        self.harness.begin()
+        self.assertFalse(self.harness.charm.state.target_created)
+        self.assertFalse(self.harness.charm.state.enable_tls)
+        self.assertEqual(self.harness.charm.state.additional_trusted_ips, [])
+
+    def add_cluster_relation(self):
+        rel_id = self.harness.add_relation('cluster', 'ceph-iscsi')
+        self.harness.add_relation_unit(
+            rel_id,
+            'ceph-iscsi/1',
+            {
+                'ingress-address': '10.0.0.2',
+                'gateway_ready': 'True',
+                'gateway_fqdn': 'ceph-iscsi-1.example'
+            }
+        )
+        return rel_id
+
+    @patch('socket.getfqdn')
+    def test_on_create_target_action(self, _getfqdn):
+        _getfqdn.return_value = 'ceph-iscsi-0.example'
+        self.add_cluster_relation()
+        self.harness.begin()
+        action_event = MagicMock()
```

Comment:
I would rather temporary mock `action_get` instead similar to `network_get` and do `self.harness.charm.on.create_target_action.emit()`.

I think it's another place where the test harness can be improved.

@jameinel ^ FYI

Reply:
Thanks for the suggestion @dshcherb , I tried the approach you suggested but it fails because ```ActionEvent().restore``` does a check of the Juju environment variable to ensure the event matches the action https://github.com/canonical/operator/blob/master/ops/charm.py#L50